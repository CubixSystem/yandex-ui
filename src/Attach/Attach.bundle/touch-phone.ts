import { withRegistry, Registry } from '@bem-react/di';
import { compose, composeU } from '@bem-react/core';

// base
import { Attach as AttachTouchPhone, cnAttach } from '../Attach@touch-phone';
// _size
import { withSizeL } from '../_size/Attach_size_l';
import { withSizeM } from '../_size/Attach_size_m';
import { withSizeS } from '../_size/Attach_size_s';

// Registry dependencies â€” Button
import { Button as ButtonDesktop } from '../../Button/Button@touch-phone';
import { withSizeL as withButtonSizeL } from '../../Button/_size/Button_size_l';
import { withSizeM as withButtonSizeM } from '../../Button/_size/Button_size_m';
import { withSizeS as withButtonSizeS } from '../../Button/_size/Button_size_s';
import { withViewDefault as withButtonViewDefault } from '../../Button/_view/Button_view_default';
import { withThemeNormal as withButtonThemeNormal } from '../../Button/_theme/Button_theme_normal';

const registry = new Registry({ id: cnAttach() });

const Button = compose(
    withButtonThemeNormal,
    withButtonViewDefault,
    composeU(withButtonSizeL, withButtonSizeM, withButtonSizeS),
)(ButtonDesktop);

registry.set('Button', Button);

export * from '../Attach@touch-phone';

export const Attach = compose(
    composeU(withSizeL, withSizeM, withSizeS),
    withRegistry(registry),
)(AttachTouchPhone);
