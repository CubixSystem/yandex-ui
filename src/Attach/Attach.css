.Attach {
    display: inline-flex;

    white-space: nowrap;
}

/* Оптимизация css для элементов holder && reset */
.Attach-IconFile,
.Attach-Reset {
    width: 1px;
    height: 1px;
}

.Attach_disabled .Attach-Holder {
    pointer-events: none;

    opacity: .6;
}

.Attach_disabled .Attach-Control {
    cursor: default;
}
