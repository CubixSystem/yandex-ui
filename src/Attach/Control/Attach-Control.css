/**
 * 1. Хак для  IE 11 для скрытия курсора — ISL-3557
 */
@media all and (min-width: 0) {
    .Attach-Control {
        position: absolute;
        top: 0;
        right: 0;

        display: block;

        width: 100%;
        height: 100%;

        cursor: pointer;

        opacity: 0;

        touch-action: manipulation;
    }

    .Attach-Control::-ms-value {
        display: none; /* 1 */
    }

    .Attach-Control::-webkit-file-upload-button {
        cursor: pointer;
    }
}
