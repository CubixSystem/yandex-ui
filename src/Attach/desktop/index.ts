export * from '../Attach@desktop';
export * from '../_size/Attach_size_l';
export * from '../_size/Attach_size_m';
export * from '../_size/Attach_size_s';
