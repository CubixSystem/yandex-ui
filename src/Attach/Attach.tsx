import React, {
    FC,
    ChangeEvent,
    MouseEvent,
    ChangeEventHandler,
    MouseEventHandler,
    RefObject,
    useState,
    useCallback,
    useRef,
    useEffect,
    KeyboardEvent,
} from 'react';
import { useComponentRegistry } from '@bem-react/di';

import { mergeRefs } from '../lib/mergeRefs';
import { Keys, isKeyCode } from '../lib/keyboard';
import { useUniqId } from '../useUniqId';
import { cnAttach, extensionsMap } from './Attach.const';
import { IAttachRegistry } from './Attach.registry/interface';
import './Attach.css';

export * from './Attach.const';

export interface IAttachProps {
    /**
     * Дополнительный className
     */
    className?: string;

    /**
     * Ссылка на DOM-элемент нативного контрола
     */
    controlRef?: RefObject<HTMLInputElement>;

    /**
     * Неактивное состояние компонента.
     * Состояние, при котором компонент отображается, но недоступен для действий пользователя
     */
    disabled?: boolean;

    /**
     * Наличие элемента `Holder`
     */
    hasHolder?: boolean;

    /**
     * Текст внутри элемента `Holder`, когда файл не выбран
     */
    holderText?: string;

    /**
     * Ширина текста внутри элемента `Holder`
     */
    holderTextWidth?: number;

    /**
     * Уникальный id компонента
     *
     * @default useUniqId('xuniq')
     */
    id?: string;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLElement>;

    /**
     * Имя компонента.
     */
    name?: string;

    /**
     * Обработчик, вызываемый при смене файла
     */
    onChange?: ChangeEventHandler<HTMLInputElement>;

    /**
     * Обработчик, вызываемый при очистке файла
     */
    onClearClick?: MouseEventHandler<HTMLElement>;

    /**
     * Размер компонента
     */
    size?: string;

    /**
     * Стилевое оформление компонента
     */
    theme?: string;

    /**
     * Внешний вид компонента
     */
    view?: string;

    /**
     * Содержимое кнопки
     */
    children: string;

    /**
     * Последовательность перехода между кнопками при нажатии на кнопку Tab
     * @default 0
     */
    tabIndex?: number;

    /**
     * Допустимые типы файлов
     */
    accept?: string;
}

type AttachState = {
    /**
     * Название файла
     */
    fileName?: string;

    /**
     * Расширение файла
     */
    fileExtension?: string;
};

/**
 * Компонент для создания кнопки выбора файла, предназначенного для отправки на сервер.
 * @param {IAttachProps} props
 */
export const Attach: FC<IAttachProps> = ({
    children,
    className,
    controlRef,
    disabled,
    hasHolder,
    holderText,
    holderTextWidth,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    id = useUniqId('xuniq'),
    innerRef,
    name,
    onChange,
    onClearClick,
    size,
    theme,
    view,
    tabIndex = 0,
    accept,
    ...props
}) => {
    const [{ fileName, fileExtension }, setFileMeta] = useState<AttachState>({});
    const { Button, Holder, Control } = useComponentRegistry<IAttachRegistry>(cnAttach());
    const controlInternalRef = useRef<HTMLInputElement>(null);

    useEffect(() => {
        mergeRefs(controlInternalRef, controlRef);
    }, [controlInternalRef, controlRef]);

    const onInternalChange = useCallback(
        (event: ChangeEvent<HTMLInputElement>) => {
            const fileName = event.target.value.split('\\').pop() as string;
            const rawFileExtension = (fileName.split('.').pop() as string).toLowerCase();
            const fileExtension = extensionsMap[rawFileExtension] || 'unknown';

            if (onChange !== undefined) {
                onChange(event);
            }

            setFileMeta({ fileName, fileExtension });
        },
        [onChange],
    );

    const onInternalClearClick = useCallback(
        (event: MouseEvent<HTMLElement>) => {
            if (controlInternalRef.current !== null) {
                if (onChange !== undefined) {
                    // Создаем синтетическое событие для эмуляции очистки контрола.
                    const syntheticEvent = Object.create(event);
                    syntheticEvent.target = controlInternalRef.current;
                    syntheticEvent.currentTarget = controlInternalRef.current;

                    controlInternalRef.current.value = '';

                    onChange(syntheticEvent);
                } else {
                    controlInternalRef.current.value = '';
                }
            }

            if (onClearClick !== undefined) {
                onClearClick(event);
            }

            setFileMeta({ fileName: undefined, fileExtension: undefined });
        },
        [onChange, onClearClick],
    );

    const onInternalKeyUp = useCallback(
        (event: KeyboardEvent<HTMLElement>) => {
            if (controlInternalRef.current !== null && isKeyCode(event.keyCode, [Keys.SPACE, Keys.ENTER])) {
                controlInternalRef.current.click();
            }
        },
        [controlInternalRef],
    );

    return (
        <span {...props} ref={innerRef} className={cnAttach({ disabled }, [className])}>
            <Button
                disabled={disabled}
                theme={theme as any}
                view={view as any}
                size={size as any}
                as="span"
                // Указываем role и aria-label для доступности.
                role="button"
                aria-label={children}
                tabIndex={disabled ? -1 : tabIndex}
                onKeyUp={onInternalKeyUp}
                className={cnAttach('Button')}
                // @ts-ignore
                autoComplete={null}
                // @ts-ignore
                type={null}
                addonAfter={
                    <Control
                        disabled={disabled}
                        id={id}
                        name={name}
                        onChange={onInternalChange}
                        innerRef={controlInternalRef}
                        accept={accept}
                    />
                }
            >
                {children}
            </Button>
            {hasHolder && (
                <Holder
                    id={id}
                    fileName={fileName}
                    fileExtension={fileExtension}
                    textWidth={holderTextWidth}
                    onClearClick={onInternalClearClick}
                >
                    {holderText}
                </Holder>
            )}
        </span>
    );
};

Attach.displayName = cnAttach();
