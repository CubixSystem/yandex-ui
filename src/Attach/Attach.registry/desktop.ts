import { Registry } from '@bem-react/di';

import { cnAttach } from '../Attach.const';
import { Holder } from '../Holder/Attach-Holder';
import { Control } from '../Control/Attach-Control';

export const attachRegistry = new Registry({ id: cnAttach() });

attachRegistry
    .set('Holder', Holder)
    .set('Control', Control)
    .set('Button', () => {
        // TODO: Использовать error api из новой версии react.
        throw new Error(`Для работы "Attach" требуется установить компонент "Button" в реестр "${cnAttach()}".`);
    });
