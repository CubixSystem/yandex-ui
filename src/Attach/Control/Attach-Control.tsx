import React, { FC, ChangeEventHandler, RefObject } from 'react';

import { cnAttach } from '../Attach.const';
import './Attach-Control.css';

export interface IControlProps {
    /**
     * ID компонента.
     */
    id: string;

    /**
     * Обработчик, вызываемый при смене файла
     */
    onChange: ChangeEventHandler<HTMLInputElement>;

    /**
     * Имя компонента
     *
     * @default 'attachment'
     */
    name?: string;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLInputElement>;

    /**
     * HTML атрибут `disabled`
     */
    disabled?: boolean;

    /**
     * Допустимые типы файлов
     */
    accept?: string;
}

export const Control: FC<IControlProps> = ({
    disabled,
    id,
    innerRef,
    name = 'attachment',
    onChange,
    accept,
    ...props
}) => (
    <input
        {...props}
        autoComplete="off"
        className={cnAttach('Control')}
        disabled={disabled}
        id={id}
        name={name}
        onChange={onChange}
        ref={innerRef}
        tabIndex={-1}
        type="file"
        accept={accept}
    />
);
