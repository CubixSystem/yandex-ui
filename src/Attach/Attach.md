# Attach

<a href='https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/components/Attach' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Исходники]/[Github
][green]/badge.svg' /></a> <a href='https://search.yandex-team.ru/stsearch?text=Attach.ts&facet.queue=ISL&facet.type=bug&facet.status=128' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Известные проблемы]/[Startrek][blue]/badge.svg' /></a>

<!-- description:start -->
Компонент для создания кнопки выбора файла, предназначенного для отправки на сервер.
<!-- description:end -->

## Пример использования

Конфигурация темы на уровне проекта:

```ts
// src/lib/theme.ts
import { configureRootTheme } from '@yandex-lego/components/Theme'
import { theme } from '@yandex-lego/components/Theme/presets/default'

configureRootTheme({ theme })
```

Использование с нужным набором модификаторов:

```ts
// src/App.ts
import React from 'react'
import { compose } from '@bem-react/core'
import { withRegistry, Registry } from '@bem-react/di'

import {
  Button as ButtonDesktop,
  withSizeM as withButtonSizeM,
  withViewDefault as withButtonViewDefault,
} from '@yandex-lego/components/Button/desktop'

import {
  cnAttach,
  Attach as AttachDesktop,
  withSizeM as withAttachSizeM,
} from '@yandex-lego/components/Attach/desktop'

const Button = compose(withButtonSizeM, withButtonViewDefault)(ButtonDesktop)

const attachRegistry = new Registry({ id: cnAttach() })
attachRegistry.set('Button', Button)

// Композиция из различных модификаторов
const Attach = compose(
  withAttachSizeM,
  withRegistry(attachRegistry),
)(AttachDesktop)

const App = () => (
  <Attach hasHolder holderText="no file chosen" view="default" size="m">
    Select file
  </Attach>
)
```

Использование с полным набором модификаторов:

```ts
// src/App.ts
import React from 'react'
import { Attach } from '@yandex-lego/components/Attach/desktop/bundle'

const App = () => (
  <Attach hasHolder holderText="no file chosen" view="default" size="m">
    Select file
  </Attach>
)
```

## Примеры

### Вид компонента

Чтобы изменить вид компонента, используйте свойство `_view` со значением как у компонента `Button`.

{{%story::desktop:controls-attach-desktop--view%}}

### Размер компонента

Чтобы изменить размер компонента, установите свойство `size` в одно из допустимых значений: `"s"`, `"m"`, `"l"`.

{{%story::desktop:controls-attach-desktop--size%}}

### Текст сообщения

Чтобы отобразить текст сообщения, установите свойство `hasHolder` в значение `true`.

{{%story::desktop:controls-attach-desktop--has-holder%}}

## Свойства

<!-- props:start -->
| Свойство         | Тип                                                    | По умолчанию         | Описание                                                                                                                |
| ---------------- | ------------------------------------------------------ | -------------------- | ----------------------------------------------------------------------------------------------------------------------- |
| className?       | `string`                                               | —                    | Дополнительный className                                                                                                |
| controlRef?      | `RefObject<HTMLInputElement>`                          | —                    | Ссылка на DOM-элемент нативного контрола                                                                                |
| disabled?        | `false \| true`                                        | —                    | Неактивное состояние компонента. Состояние, при котором компонент отображается, но недоступен для действий пользователя |
| hasHolder?       | `false \| true`                                        | —                    | Наличие элемента `Holder`                                                                                               |
| holderText?      | `string`                                               | —                    | Текст внутри элемента `Holder`, когда файл не выбран                                                                    |
| holderTextWidth? | `number`                                               | —                    | Ширина текста внутри элемента `Holder`                                                                                  |
| id?              | `string`                                               | `useUniqId('xuniq')` | Уникальный id компонента                                                                                                |
| innerRef?        | `RefObject<HTMLElement>`                               | —                    | Ссылка на корневой DOM-элемент компонента                                                                               |
| name?            | `string`                                               | —                    | Имя компонента.                                                                                                         |
| onChange?        | `(event: ChangeEvent<HTMLInputElement>) => void`       | —                    | Обработчик, вызываемый при смене файла                                                                                  |
| onClearClick?    | `(event: MouseEvent<HTMLElement, MouseEvent>) => void` | —                    | Обработчик, вызываемый при очистке файла                                                                                |
| size?            | `string`                                               | —                    | Размер компонента                                                                                                       |
| theme?           | `string`                                               | —                    | Стилевое оформление компонента                                                                                          |
| view?            | `string`                                               | —                    | Внешний вид компонента                                                                                                  |
| children         | `string`                                               | —                    | Содержимое кнопки                                                                                                       |
| tabIndex?        | `number`                                               | `0`                  | Последовательность перехода между кнопками при нажатии на кнопку Tab                                                    |
| accept?          | `string`                                               | —                    | Допустимые типы файлов                                                                                                  |
<!-- props:end -->
