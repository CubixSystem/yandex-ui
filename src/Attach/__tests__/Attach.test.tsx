import React, { createRef } from 'react';
import { withRegistry, Registry } from '@bem-react/di';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { delay } from '../../internal/utils';
import { Control } from '../Control/Attach-Control';
import { cnAttach, Attach as AttachCommon } from '../Attach';
import { Attach as AttachDesktop } from '../desktop';
import { Attach as AttachTouchPad } from '../touch-pad';
import { Attach as AttachTouchPhone } from '../touch-phone';
import { Button as ButtonDesktop } from '../../Button/desktop';
import { Button as ButtonTouchPad } from '../../Button/touch-pad';
import { Button as ButtonTouchPhone } from '../../Button/touch-phone';

const attachDesktopRegistry = new Registry({ id: cnAttach() });
const attachTouchPadRegistry = new Registry({ id: cnAttach() });
const attachTouchPhoneRegistry = new Registry({ id: cnAttach() });

const ControlWithTestId = (props: any) => {
    return <Control {...props} data-testid="control" />;
};

attachDesktopRegistry.set('Button', ButtonDesktop);
attachDesktopRegistry.set('Control', ControlWithTestId);
attachTouchPadRegistry.set('Button', ButtonTouchPad);
attachTouchPadRegistry.set('Control', ControlWithTestId);
attachTouchPhoneRegistry.set('Button', ButtonTouchPhone);
attachTouchPhoneRegistry.set('Control', ControlWithTestId);

const platforms = [
    ['desktop', withRegistry(attachDesktopRegistry)(AttachDesktop)],
    ['touch-pad', withRegistry(attachTouchPadRegistry)(AttachTouchPad)],
    ['touch-phone', withRegistry(attachTouchPhoneRegistry)(AttachTouchPhone)],
] as const;

describe.each(platforms)('Attach@%s', (_platform, Attach) => {
    const render = createClientRender();

    function triggerChange(element: Element, value: string) {
        // NOTE: Если прокидывать value в событии, то будет ошибка:
        // InvalidStateError: This input element accepts a filename,
        // which may only be programmatically set to the empty string
        Object.defineProperty(element, 'value', {
            value,
            writable: true,
        });

        fireEvent.change(element);
    }

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Attach hasHolder holderText="файл не выбран" id="100500" className="mix">
                Выбрать файл
            </Attach>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(AttachCommon.displayName).toBe('Attach');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLElement>();
        render(<Attach innerRef={innerRef}>Выбрать файл</Attach>);

        await delay(100);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать ссылку на нативный DOM инпут', async () => {
        const controlRef = createRef<HTMLInputElement>();
        render(<Attach controlRef={controlRef}>Выбрать файл</Attach>);

        await delay(100);

        expect(controlRef.current).not.toBe(null);
    });

    test('должен устанавливать текст в элементах Button-Text и Attach-Text', () => {
        render(
            <Attach hasHolder holderText="Holder text">
                Button text
            </Attach>,
        );

        expect(screen.queryByText('Button text')).toBeInTheDocument();
        expect(screen.queryByText('Holder text')).toBeInTheDocument();
    });

    test('должен устанавливать ширину текста в элементе Attach-Text', () => {
        render(
            <Attach hasHolder holderText="файл не выбран" holderTextWidth={40}>
                Выбрать файл
            </Attach>,
        );

        triggerChange(screen.getByTestId('control'), 'file.png');

        expect(screen.getByText('file.png')).toHaveStyle({
            width: '40px',
        });
    });

    test('должен вызывать событие onChange с event при выборе файла', () => {
        // TODO: при переходе на 17 реакт `e.persist` убрать
        const onChange = jest.fn((e) => e.persist());
        render(<Attach onChange={onChange}>Выбрать файл</Attach>);

        triggerChange(screen.getByTestId('control'), 'file.png');

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange.mock.calls[0][0].target.value).toBe('file.png');
    });

    test('должен вызывать событие onClearClick & onChange при клике в reset', () => {
        const onChange = jest.fn();
        const onClearClick = jest.fn();

        const { setProps } = render(
            <Attach hasHolder holderText="файл не выбран" onClearClick={onClearClick}>
                Выбрать файл
            </Attach>,
        );

        triggerChange(screen.getByTestId('control'), 'file.png');
        setProps({ onChange });

        // TODO: Так делать не стоит, но выбора нет. Нужно рефакторить сами компоненты
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const reset = document.querySelector('.Attach-Reset')!;
        fireEvent.click(reset);

        expect(onClearClick).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange.mock.calls[0][0].target.value).toBe('');
    });

    test('должен устанавливать свойство accept', () => {
        const acceptType = 'image/*';
        render(<Attach accept={acceptType}>Выбрать файл</Attach>);

        expect(screen.getByTestId('control')).toHaveAttribute('accept', acceptType);
    });
});
