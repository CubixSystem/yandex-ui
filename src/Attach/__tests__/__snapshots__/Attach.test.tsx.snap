// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attach@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Attach mix"
>
  <span
    aria-label="Выбрать файл"
    class="Button2 Attach-Button"
    role="button"
    tabindex="0"
  >
    <span
      class="Button2-Text"
    >
      Выбрать файл
    </span>
    <input
      autocomplete="off"
      class="Attach-Control"
      data-testid="control"
      id="100500"
      name="attachment"
      tabindex="-1"
      type="file"
    />
  </span>
  <span
    aria-hidden="true"
    class="Attach-Holder"
  >
    <label
      class="Attach-Text"
      for="100500"
    >
      файл не выбран
    </label>
  </span>
</span>
`;

exports[`Attach@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Attach mix"
>
  <span
    aria-label="Выбрать файл"
    class="Button2 Attach-Button"
    role="button"
    tabindex="0"
  >
    <span
      class="Button2-Text"
    >
      Выбрать файл
    </span>
    <input
      autocomplete="off"
      class="Attach-Control"
      data-testid="control"
      id="100500"
      name="attachment"
      tabindex="-1"
      type="file"
    />
  </span>
  <span
    aria-hidden="true"
    class="Attach-Holder"
  >
    <label
      class="Attach-Text"
      for="100500"
    >
      файл не выбран
    </label>
  </span>
</span>
`;

exports[`Attach@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Attach mix"
>
  <span
    aria-label="Выбрать файл"
    class="Button2 Attach-Button"
    role="button"
    tabindex="0"
  >
    <span
      class="Button2-Text"
    >
      Выбрать файл
    </span>
    <input
      autocomplete="off"
      class="Attach-Control"
      data-testid="control"
      id="100500"
      name="attachment"
      tabindex="-1"
      type="file"
    />
  </span>
  <span
    aria-hidden="true"
    class="Attach-Holder"
  >
    <label
      class="Attach-Text"
      for="100500"
    >
      файл не выбран
    </label>
  </span>
</span>
`;
